From: B Horn <b@horn.uk>
Date: Sun, 12 May 2024 11:08:23 +0100
Subject: commands/ls: Fix NULL dereference
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

The grub_strrchr() may return NULL when the dirname do not contain "/".
This can happen on broken filesystems.

Reported-by: B Horn <b@horn.uk>
Signed-off-by: B Horn <b@horn.uk>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>
(cherry picked from commit 0bf56bce47489c059e50e61a3db7f682d8c44b56)
FG: adapt context, 2.06 doesn't yet have the bug fix
6889c67fe7f2d6bbaa1f6050d9e29e616740936b that changed the variable where the
dirname is stored.
Signed-off-by: Fabian Grünbichler <f.gruenbichler@proxmox.com>
---
 grub-core/commands/ls.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/grub-core/commands/ls.c b/grub-core/commands/ls.c
index 8e98c73..04df752 100644
--- a/grub-core/commands/ls.c
+++ b/grub-core/commands/ls.c
@@ -241,9 +241,13 @@ grub_ls_list_files (char *dirname, int longlist, int all, int human)
 
 	  grub_file_close (file);
 
-	  p = grub_strrchr (dirname, '/') + 1;
+	  p = grub_strrchr (dirname, '/');
+	  if (p == NULL)
+	    goto fail;
+	  ++p;
+
 	  dirname = grub_strndup (dirname, p - dirname);
-	  if (! dirname)
+	  if (dirname == NULL)
 	    goto fail;
 
 	  all = 1;
